strings/base_collections_base.h

```diff
@@ -142,40 +235,59 @@ WINRT_EXPORT namespace winrt
+                {
+                    throw hresult_out_of_bounds();
+                }
 
-            return static_cast<D const&>(*this).unwrap_value(*std::next(static_cast<D const&>(*this).get_container().begin(), index));
+                return static_cast<D const&>(*this).unwrap_value(*std::next(static_cast<D const&>(*this).get_container().begin(), index));
+            });
         }
 
         uint32_t Size() const noexcept
         {
-            return static_cast<uint32_t>(std::distance(static_cast<D const&>(*this).get_container().begin(), static_cast<D const&>(*this).get_container().end()));
+            return static_cast<D const&>(*this).perform_shared([&]
+            {
+                return container_size();
+            });
         }
 
         bool IndexOf(T const& value, uint32_t& index) const noexcept
         {
-            auto first = std::find_if(static_cast<D const&>(*this).get_container().begin(), static_cast<D const&>(*this).get_container().end(), [&](auto&& match)
+            return static_cast<D const&>(*this).perform_shared([&]
             {
-                return value == static_cast<D const&>(*this).unwrap_value(match);
-            });
+                auto first = std::find_if(static_cast<D const&>(*this).get_container().begin(), static_cast<D const&>(*this).get_container().end(), [&](auto&& match)
+                {
+                    return value == static_cast<D const&>(*this).unwrap_value(match);
+                });
 
-            index = static_cast<uint32_t>(first - static_cast<D const&>(*this).get_container().begin());
-            return index < Size();
+                index = static_cast<uint32_t>(first - static_cast<D const&>(*this).get_container().begin());
+                return index < container_size();
+            });
         }
 
         uint32_t GetMany(uint32_t const startIndex, array_view<T> values) const
         {
-            if (startIndex >= Size())
+            return static_cast<D const&>(*this).perform_shared([&]() -> uint32_t
             {
-                return 0;
-            }
+                if (startIndex >= container_size())
+                {
+                    return 0;
+                }
+
+                uint32_t const actual = (std::min)(container_size() - startIndex, values.size());
+                this->copy_n(static_cast<D const&>(*this).get_container().begin() + startIndex, actual, values.begin());
+                return actual;
+            });
+        }
 
-            uint32_t const actual = (std::min)(Size() - startIndex, values.size());
-            this->copy_n(static_cast<D const&>(*this).get_container().begin() + startIndex, actual, values.begin());
-            return actual;
+    private:
+
+        uint32_t container_size() const noexcept
+        {
+            return static_cast<uint32_t>(std::distance(static_cast<D const&>(*this).get_container().begin(), static_cast<D const&>(*this).get_container().end()));
```

Comment:
We seem inconsistent about the requirements we place on the container. Here, we only assume that it has `begin()` and `end()`, but make no assumptions about `size()`. But `vector_base` assumes that `size()` exists. I wonder why?

Reply:
🤷‍♂️

Reply:
Views can be constructed from iterator pairs and there's no `size` on `winrt::impl::range_container`.

Reply:
Makes sense. On a side note, I've been toying with the thought of introducing C++20 concepts to C++/WinRT in few places. The collection-related code, in particular, is an area that would benefit.

Reply:
Yes, there are a few C++20 features I'd like to adopt including span and coroutines #676 - we should think about forking cppwinrt with a servicing C++17 branch and start work on C++20 features in master. Just a little nervous about how long it will take for C++20 to become mainstream since Visual C++ doesn't even have a /std:c++20 flag yet.

Reply:
Agreed. At least with span, concepts, char8_t, consteval, chrono updates, and otheres, there are clear-cut feature test macros. Coroutine support has the potential to be messier.